warehouse/warehouse.go

```diff
@@ -486,10 +491,33 @@ func uploadFrequencyExceeded(warehouse warehouseutils.WarehouseT, syncFrequency
 	return false
 }
 
+func getLastStagingFileCreatedAt(warehouse warehouseutils.WarehouseT) time.Time {
+	stmt := fmt.Sprintf(`SELECT created_at from %s WHERE source_id='%s' AND destination_id='%s' ORDER BY id DESC LIMIT 1`, warehouseutils.WarehouseStagingFilesTable, warehouse.Source.ID, warehouse.Destination.ID)
+	var createdAt sql.NullTime
+	err := dbHandle.QueryRow(stmt).Scan(&createdAt)
+	if err != nil {
+		pkgLogger.Errorf(`Error in getLastStagingFileCreatedAt: %v`, err)
+		return time.Now()
+	}
+	return createdAt.Time
+}
+
 func setLastEventTime(warehouse warehouseutils.WarehouseT, lastEventAt time.Time) {
+	unixTime := lastEventAt.Unix()
+
+	freqInS := getUploadFreqInS(warehouseutils.GetConfigValue(warehouseutils.SyncFrequency, warehouse))
+
+	lastStagingFileCreatedAt := getLastStagingFileCreatedAt(warehouse)
+	// set lastEvent to currentTime in cases where
+	// 1. if there is lag in creating staging files - last staging file is more than freqInS old
+	// 2. and if there is less than freqInS worth of data in pending staging files
+	if time.Now().Add(time.Duration(-freqInS)*time.Second).Sub(lastStagingFileCreatedAt) > 0 && lastStagingFileCreatedAt.Sub(lastEventAt) < time.Duration(freqInS)*time.Second {
+		unixTime = time.Now().Unix()
```

Comment:
This fixes our issue, but the naming is kind of messed up. Basically, in the same map, we are storing last_exec_at when there is lag and last_event_at otherwise.

Instead lets keep setting last_exec_at like we used to do before because it only has in-memory existence. And then fix the check accounting for the lag in `uploadFrequencyExceeded`?